### PR TITLE
Integration4 for Snapdragon Flight

### PR DESCRIPTION
**Pull request by julianoes**
This is a rebase of integration3 on current master (570fb97163e4a9c798c9f5d9db2109d874903879).

On top of that, the branch contains several merge conflict fixes, as well as build fixes and ifdefs in order not to break existing functionality.

Currently, the following build:

- [x] ```make px4fmu-v2_default```
- [x] ```make posix gazebo```
- [x] ```make posix jmavsim```
- [x] ```make posix_rpi2_default```
- [ ] ```make posix_rpi2_release```
- [x] ```make posix_eagle_default```
- [x] ```make qurt_eagle_default```

The following are flight tested:

- [x] ```make px4fmu-v2_default``` (**TODO before merging**)
- [x] ```make posix gazebo```
- [x] ```make posix jmavsim```
- [ ] ```make posix_rpi2_default``` (still needs sensor integration, won't fly just yet)
- [ ] ```make posix_rpi2_release``` (still needs sensor integration, won't fly just yet)
- [ ] ```make posix_eagle_default``` (WIP)
- [ ] ```make qurt_eagle_default``` (WIP)

@mcharleb and @LorenzMeier please review and check for merge mistakes (especially removed/added files) and check for things that could potentially flying with NuttX.

@tumbili (or anyone else) please do a regression flight test on hardware running NuttX.

~~Also, before merging this, we should try to include @jywilson's fixes as well.~~ Done

This pull request replaces #3723.

### CODE REVIEW

**Comment by mcharleb** [comment]
@julianoes posix_apps.py and qurt_apps.py no longer needed?

File: src/modules/commander/commander.cpp [hunk 1]
```diff
@@ -2188,8 +2192,13 @@ int commander_thread_main(int argc, char *argv[])
 		}
 
 		/* RC input check */
+#ifndef __PX4_QURT
 		if (!status.rc_input_blocked && sp_man.timestamp != 0 &&
 		    (hrt_absolute_time() < sp_man.timestamp + (uint64_t)(rc_loss_timeout * 1e6f))) {
+#else
```
Comment (mcharleb) [inline src/modules/commander/commander.cpp]: @jywilson Is there a proper fix for this?

**Comment by bo-rc** [comment]
Hi @julianoes , when you say the target `posix_rpi2_default` "still needs sensor integration" Do you mean that the the DriverFramework needs some work to make `posix_rpi2_default` flyable? 

For example,  i saw the 
```cpp
int SPIDevObj::_readReg(uint8_t address, uint8_t &val);
```
 has implementations for `__QURT` only.

If the implementations of relevant sensor accessing functions for raspberry pi2 are also done in a similar style as `__QURT`, we can go ahead to do test-flights with `posix_rpi2_default`?

**Comment by julianoes** [comment]
@mcharleb posix_apps.py and qurt_apps.py must have gotten lost in the merge. I'll add them back. Thanks for checking.

**Comment by julianoes** [comment]
@bo-rc Yes that's pretty much right.

**Comment by julianoes** [comment]
@LorenzMeier any hints why travis fails?
All I see is: ```Done. Your build exited with 1.```

**Comment by julianoes** [comment]
@mcharleb See 7aeef19d917e5063ff390c2d9fa6145975b02d0b: I've revived the two scripts, however I have no idea what they are for and how they are used. Please point me to directions or add comments in the files.

**Comment by mhkabir** [comment]
@julianoes You forgot the ```posix_rpi2_release``` build. That's the one we're most interested in, since it cross-compiles for RPi2. I've added it to the OP.

**Comment by LorenzMeier** [comment]
Let's strip the release config for RPI2 entirely. There is no reason to have it. The existence of the snapdragon configs is solely because there are open source and proprietary drivers and we want to support them both in the interim.

**Comment by bo-rc** [comment]
@julianoes Thanks!! @mhkabir are you already working on `posix_rpi2_default` sensor integration in the DriverFramework project? If no one is working on it yet, i would like to take it.

**Comment by mhkabir** [comment]
Hmm. So should we just crosscompile on the default config?

The way it is currently leaves an option to compile on the Pi (default) and
cross compile (release)

**Comment by LorenzMeier** [comment]
Ok, now I understand the difference - is there a real use case for compiling on the Pi itself?

**Comment by bo-rc** [comment]
Hi,

I would prefer to keep the native build option on the Pi itself. so that
users can just ssh to the Pi from any computer (maybe a Windows, or even a
tablet on the field) and change some code then compile.

My use case is ssh to my Pi at home from my office terminal (no tool chain
installed) and compile some code to run on the Pi.

The compilation time cost on my RPi2 is good enough so i like it.

**Comment by julianoes** [comment]
Ok, I'll try out if the native compilation still works.
I didn't understand the difference between default and release. What about renaming them to cross and native?

**Comment by LorenzMeier** [comment]
Yes, renaming them makes a lot of sense.

**Comment by LorenzMeier** [comment]
@bo-rc It would be great if you could start with the RPI2 sensors. We really would like to push forward. I will bring this branch on master today with Julian.

**Comment by julianoes** [comment]
I pushed another commit that prevents the crashes due to the I2C write troubles (see https://github.com/PX4/DriverFramework/issues/23).
Let's wait for the travis checks and do a quick NuttX flight in the meantime.

**Comment by LorenzMeier** [comment]
It flies well on Pixhawk / NuttX. However, on my bench Pixhawk the artificial horizon is not updating any more. I need to work out if that is a QGC issue or an issue of this branch:
https://github.com/mavlink/qgroundcontrol/issues/2851

**Comment by NaterGator** [comment]
@LorenzMeier pretty sure it is QGC, I am having that issue too.

**Comment by julianoes** [comment]
@LorenzMeier ok thanks for testing, I can rebase this to master if you want now.

**Comment by LorenzMeier** [comment]
Please rebase, yes.

**Comment by bo-rc** [comment]
@LorenzMeier yes i am working on RPi2 sensors.

**Comment by julianoes** [comment]
Rebased to master, closing.

**Pull request closed without merging**